### PR TITLE
Keycloak 15700

### PR DESCRIPTION
**Pull request by briandooley**

### CODE REVIEW

**Review by andymunro — APPROVED** [review]

**Comment by andymunro** [comment]
This one still has a few conflicting files.

File: server_admin/topics/admin-console-permissions/master-realm.adoc [hunk 1]
```diff
@@ -1,29 +1,23 @@
...
-are allowed to create new realms.  They will be granted full access to any new realm they create.
+Users with the `admin` role are super users and have access to manage any realm on the server. Users with the `create-realm` role can create new realms and these users have full access to any new realm they create.
 
 ==== Realm Specific Roles
 
-Admin users within the `master` realm can be granted management privileges to one or more other realms in the system.
-Each realm in {project_name} is represented by a client in the `master` realm.
-The name of the client is `<realm name>-realm`.  These clients each have client-level roles defined which define varying
-level of access to manage an individual realm.
+It is possible to grant management privileges, for multiple realms in the system, to admin users within the `master` realm.
```
Comment (andymunro) [inline server_admin/topics/admin-console-permissions/master-realm.adoc]: Change "It is possible to" to "You can"

File: server_admin/topics/admin-console-permissions/master-realm.adoc [hunk 2]
```diff
@@ -40,9 +34,9 @@ The roles available are:
 * manage-identity-providers
 * impersonation
 
-Assign the roles you want to your users and they will only be able to use that specific part of the administration console.
-
-IMPORTANT: Admins with the `manage-users` role will only be able to assign admin roles to users that they themselves have.  So, if an admin has the `manage-users` role but doesn't have the `manage-realm` role, they will not be able to assign this role.
-
-
+Users with these roles can use that specific part of the administration console only.
```
Comment (andymunro) [inline server_admin/topics/admin-console-permissions/master-realm.adoc]: Change "administration console" to "Admin Console"

File: server_admin/topics/admin-console-permissions/master-realm.adoc [hunk 3]
```diff
@@ -40,9 +34,9 @@ The roles available are:
...
 
+[NOTE]
+====
+Admins with the `manage-users` role can assign admin roles to users that only admins with the `manage-users` role have. For example, if an admin has the `manage-users` role but does not have the `manage-realm` role, they will be unable to assign the `manage-realm` role to other users.
```
Comment (andymunro) [inline server_admin/topics/admin-console-permissions/master-realm.adoc]: Refer to a user with administrator rights as an “administrator,” not an admin.

File: server_admin/topics/clients/oidc/service-accounts.adoc [hunk 4]
```diff
@@ -0,0 +1,57 @@
+[[_service_accounts]]
+
+==== Service Accounts
+
+Each OIDC client has a built-in _service account_ which allows it to obtain an access token.
+This is covered in the OAuth 2.0 specifiation under <<_client_credentials_grant,Client Credentials Grant>>.
```
Comment (andymunro) [inline server_admin/topics/clients/oidc/service-accounts.adoc:6]: This capability

File: server_admin/topics/clients/oidc/service-accounts.adoc [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+To use this feature you must set the <<_access-type, Access Type>> of your client to `confidential`.  When you do this,
+the `Service Accounts Enabled` switch is displayed.  You need to toggle this switch to ON.  Also make sure that you have
```
Comment (andymunro) [inline server_admin/topics/clients/oidc/service-accounts.adoc:8]: Remote "You need to"

File: server_admin/topics/realms/keys.adoc [hunk 6]
```diff
@@ -1,94 +1,124 @@
+
 [[realm_keys]]
 === Realm Keys
 
-The authentication protocols that are used by {project_name} require cryptographic signatures and sometimes
-encryption.  {project_name} uses asymmetric key pairs, a private and public key, to accomplish this.
+{project_name}'s authentication protocols require cryptographic signatures and, in some cases, encryption.  {project_name} implements asymmetric key pairs (a private and public key) to do this.
```
Comment (andymunro) [inline server_admin/topics/realms/keys.adoc]: change "to do this" to "address this situation."

File: server_admin/topics/realms/keys.adoc [hunk 7]
```diff
@@ -1,94 +1,124 @@
...
+To view the active keys for a realm, perform the following:
 
-To view the active keys for a realm select the realm in the admin console click on `Realm settings` then `Keys`. This
-will show the currently active keys for the realm.
+.Procedure
+. Select the realm in the admin console.
+. Click `Realm settings`.
+. Click `Keys`. 
+. Click `Passive` to view passive keys.
+. Click `Disabled` to view disabled keys.
```
Comment (andymunro) [inline server_admin/topics/realms/keys.adoc]: Asterisks no backticks.

File: server_admin/topics/realms/keys.adoc [hunk 8]
```diff
@@ -1,94 +1,124 @@
...
-It's recommended to regularly rotate keys. To do so you should start by creating new keys with a higher priority than
-the existing active keys. Or create new keys with the same priority and making the previous keys passive.
+Rotate keys regularly by:
 
-Once new keys are available all new tokens and cookies will be signed with the new keys. When a user authenticates to an
-application the SSO cookie is updated with the new signature. When OpenID Connect tokens are refreshed new tokens are
-signed with the new keys. This means that over time all cookies and tokens will use the new keys and after a while the
-old keys can be removed.
+* Creating new keys with a higher priority than the existing active keys or
+* Create new keys with the same priority and make the previous keys passive.
```
Comment (andymunro) [inline server_admin/topics/realms/keys.adoc]: Make it "Creating" and "making"

File: server_admin/topics/realms/keys.adoc [hunk 9]
```diff
@@ -1,94 +1,124 @@
...
-supported values are 1024, 2048 and 4096).
+.Procedure
+. Select the realm in the admin console.
+. Click `Realm settings`.
+. Click the `Keys` tab.
+. Click the 'Providers' tab.
+. Click `Add keystore` and select `rsa-generated`. 
+. Enter a number in the `Priority` field. This number determines if the new key pair becomes the active key pair.
+. Select a value for `keysize`.
+. Click *Save*.
```
Comment (andymunro) [inline server_admin/topics/realms/keys.adoc]: Asterisks

File: server_admin/topics/realms/keys.adoc [hunk 10]
```diff
@@ -1,94 +1,124 @@
...
+.Procedure
+. Select the realm in the admin console.
+. Click `Realm settings`.
+. Click the `Keys` tab.
+. Click the 'Providers' tab.
+. Click `Add keystore` and select `rsa`. 
+. Enter a number in the `Priority` field. This number determines if the new key pair becomes the active key pair.
+. Click *Select file* beside `Private RSA Key` to upload the private key file.
+. If you have a signed certificate for your private key,  click *Select file* beside `X509 Certificate` to upload the certificate file. {project_name} autmatically generates a self-signed certificate if you do not upload a certificate.
+. Click *Save*.
```
Comment (andymunro) [inline server_admin/topics/realms/keys.adoc]: Asterisks

File: server_admin/topics/realms/keys.adoc [hunk 11]
```diff
@@ -1,94 +1,124 @@
...
+. Click `Realm settings`.
+. Click the `Keys` tab.
+. Click the 'Providers' tab.
+. Click `Add keystore` and select `java-keystore`. 
+. Enter a number in the `Priority` field. This number determines if the new key pair becomes the active key pair.
+. Enter a value for `Keystore`.
+. Enter a value for `Keystore Password`.
+. Enter a value for `Key Alias`.
+. Enter a value for `Key Password`.
+. Click *Save*.
```
Comment (andymunro) [inline server_admin/topics/realms/keys.adoc]: Asterisks

File: server_admin/topics/realms/keys.adoc [hunk 12]
```diff
@@ -1,94 +1,124 @@
...
 ==== Making keys passive
 
-Locate the keypair in `Active` then click on the provider in the `Provider` column. This will take you to the
-configuration screen for the key provider for the keys. Click on `Active` to turn it `OFF`, then click on `Save`. The
-keys will no longer be active and can only be used for verifying signatures.
+.Procedure
+. Select the realm in the admin console.
+. Click `Realm settings`.
+. Click the `Keys` tab.
+. Click the 'Active' tab.
```
Comment (andymunro) [inline server_admin/topics/realms/keys.adoc]: Asterisks

File: server_admin/topics/realms/keys.adoc [hunk 13]
```diff
@@ -1,94 +1,124 @@
...
-Locate the keypair in `Active` then click on the provider in the `Provider` column. This will take you to the
-configuration screen for the key provider for the keys. Click on `Enabled` to turn it `OFF`, then click on `Save`. The
-keys will no longer be enabled.
-
-Alternatively, you can delete the provider from the `Providers` table.
+.Procedure
+. Select the realm in the admin console.
+. Click `Realm settings`.
+. Click the `Keys` tab.
+. Click the 'Active' tab.
```
Comment (andymunro) [inline server_admin/topics/realms/keys.adoc]: Asterisks

File: server_admin/topics/realms/keys.adoc [hunk 14]
```diff
@@ -1,94 +1,124 @@
...
-entities. However if you think that your realm signing key was compromised, you should first generate new keypair as described above and
-then immediately remove the compromised keypair.
+{project_name} has its signing keys stored locally. The signing keys are never shared with client applications, users or other entities. If your realm signing key is compromised,  <<adding-a-generated-key-pair, generate a new key>> and then immediately remove the compromised key.
+
+To ensure that client applications do not accept the tokens signed by compromised keys, update and push the not-before policy for the realm.
+
+.Procedure
+. Click *Clients* in the menu.
+. Click `security-admin-console`.
+. Click the `Revocation` tab.
```
Comment (andymunro) [inline server_admin/topics/realms/keys.adoc]: Asterisks

**Review by andymunro — CHANGES_REQUESTED** [review]
Thanks, Brian.  A few changes.

Also, I am not sure why the "service-accounts" file is in there, but I don't think it matter since we have a new file proc-using-a-service account.

**Comment by andymunro** [comment]
Okay, I found the changes.  Looks good.  Will merge it when I am ready to integrate this content.

**Review by andymunro — APPROVED** [review]

**Comment by briandooley** [comment]
@andymunro Ok to merge this?

**Comment by andymunro** [comment]
Will merge it today @briandooley

**Comment by andymunro** [comment]
Merged. Thanks @briandooley

**Pull request merged**